File: 2020/rs/src/bin/day_11.rs
use std::fs;

fn in_bounds(grid: &Vec<Vec<char>>, y: i32, x: i32) -> bool {
    0 <= y && 0 <= x && (y as usize) < grid.len() && (x as usize) < grid[0].len()
}

fn neighbors(grid: &Vec<Vec<char>>, y: i32, x: i32) -> u32 {
    let mut count = 0;
    for &y1 in &[y-1, y, y+1] {
        for &x1 in &[x-1, x, x+1] {
            if in_bounds(grid, y1, x1) && (y1 != y || x1 != x) && grid[y1 as usize][x1 as usize] == '#' {
                count += 1
            }
        }
    }
    count
}

fn visible(grid: &Vec<Vec<char>>, y: i32, x: i32) -> u32 {
    let mut count = 0;
    for &dy in &[-1, 0, 1] {
        for &dx in &[-1, 0, 1] {
            if dy == 0 && dx == 0 {
                continue;
            }
            let (mut y1, mut x1) = (y, x);
            loop {
                y1 += dy;
                x1 += dx;
                if !in_bounds(grid, y1, x1) || grid[y1 as usize][x1 as usize] == 'L' {
                    break;
                }
                if grid[y1 as usize][x1 as usize] == '#' {
                    count += 1;
                    break;
                }
            }
        }
    }
    count
}

fn simulate(mut grid: Vec<Vec<char>>, advanced: bool) -> usize {
    let mut stable = false;
    while !stable {
        stable = true;
        let mut g_prime = grid.clone();
        for y in 0..grid.len() {
            for x in 0..grid[0].len() {
                let n = if advanced { visible(&grid, y as i32, x as i32) } else { neighbors(&grid, y as i32, x as i32) };
                if n == 0 && grid[y][x] == 'L' {
                    g_prime[y][x] = '#';
                    stable = false;
                } else if n >= 4 + (advanced as u32) && grid[y][x] == '#' {
                    g_prime[y][x] = 'L';
                    stable = false;
                }
            }
        }
        grid = g_prime;
    }
    grid.iter().flatten().filter(|&&s| s == '#').count()
}

fn main() {
    let input = fs::read_to_string("../input/input_11.txt").unwrap();
    let grid: Vec<Vec<char>> = input.lines().map(|s| s.chars().collect()).collect();
    println!("Part 1: {}", simulate(grid.clone(), false));
    println!("Part 2: {}", simulate(grid, true));
}

File: python/2020/day_11.py
import itertools as it
from copy import deepcopy
from pathlib import Path


def in_bounds(grid, y, x) -> bool:
    return 0 <= y < len(grid) and 0 <= x < len(grid[0])

def neighbors(grid, y, x) -> int:
    count = 0
    for y1, x1 in it.product((y-1, y, y+1), (x-1, x, x+1)):
        if in_bounds(grid, y1, x1) and (y1 != y or x1 != x) and grid[y1][x1] == '#':
            count += 1
    return count

def visible(grid, y, x) -> int:
    count = 0
    for dy, dx in it.product((-1, 0, 1), repeat=2):
        if dy == dx == 0:
            continue
        y1, x1 = y, x
        while True:
            y1, x1 = y1 + dy, x1 + dx
            if not in_bounds(grid, y1, x1) or grid[y1][x1] == 'L':
                break
            if grid[y1][x1] == '#':
                count += 1
                break
    return count


def simulate(grid, advanced) -> int:
    grid = deepcopy(grid)
    stable = False
    while not stable:
        stable = True
        g_prime = deepcopy(grid)
        for y, x in it.product(range(len(grid)), range(len(grid[0]))):
            n = visible(grid, y, x) if advanced else neighbors(grid, y, x)
            if grid[y][x] == 'L' and n == 0:
                g_prime[y][x] = '#'
                stable = False
            elif grid[y][x] == '#' and n >= 4 + advanced:
                g_prime[y][x] = 'L'
                stable = False
        grid = g_prime
    return sum(seat == '#' for row in grid for seat in row)

def main():
    text = (Path(__file__).parent / "../../input/2020/input_11.txt").read_text()
    grid = [list(line) for line in text.splitlines()]
    print("Part 1: ", simulate(grid, False))
    print("Part 2: ", simulate(grid, True))


if __name__ == "__main__":
    main()
